### PR TITLE
Replaced all instances of outdated type annotations (List, Dict, etc.…)

### PR DESCRIPTION
**Pull request by CulmoneY**
## Proposed Changes

_Updated the PythonTA codebase to use modern type annotations. Specifically, I replaced `List` with `list`, `Dict` with `dict`, `Set` with `set`, and `Tuple` with `tuple`. I addressed the following two requirements:_

1. _Added `from __future__ import annotations` to the top of each file where these type annotations are used, if it wasn't present already._
2. _Removed any unused imports of `List`, `Dict`, `Set`, and `Tuple` to clean up the codebase._

## Type of Change

_(Write an `X` or a brief description next to the type or types that best describe your changes.)_

| Type                                                                                    | Applies? |
| --------------------------------------------------------------------------------------- | -------- |
| 🚨 _Breaking change_ (fix or feature that would cause existing functionality to change) |          |
| ✨ _New feature_ (non-breaking change that adds functionality)                          |          |
| 🐛 _Bug fix_ (non-breaking change that fixes an issue)                                  |          |
| ♻️ _Refactoring_ (internal change to codebase, without changing functionality)          |    X     |
| 🚦 _Test update_ (change that _only_ adds or modifies tests)                            |          |
| 📚 _Documentation update_ (change that _only_ updates documentation)                    |          |
| 📦 _Dependency update_ (change that updates a dependency)                               |          |
| 🔧 _Internal_ (change that _only_ affects developers or continuous integration)         |          |

## Checklist

_(Complete each of the following items for your pull request. Indicate that you have completed an item by changing the `[ ]` into a `[x]` in the raw text, or by clicking on the checkbox in the rendered description on GitHub.)_

Before opening your pull request:

- [x] I have performed a self-review of my changes.
  - Check that all changed files included in this pull request are intentional changes.
  - Check that all changes are relevant to the purpose of this pull request, as described above.
- [x] I have added tests for my changes, if applicable.
  - This is **required** for all bug fixes and new features.
- [x] I have updated the project documentation, if applicable.
  - This is **required** for new features.
- [x] I have updated the project Changelog (this is required for all changes).
- [x] If this is my first contribution, I have added myself to the list of contributors.

After opening your pull request:

- [x] I have verified that the pre-commit.ci checks have passed.
- [x] I have verified that the CI tests have passed.
- [x] I have reviewed the test coverage changes reported by Coveralls.
- [x] I have [requested a review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) from a project maintainer.

### CODE REVIEW

File: examples/ending_locations/subscript.py [hunk 1]
```diff
@@ -16,8 +17,7 @@
 x = m[  :  ]
 
 
-from typing import List
```
Comment (david-yz-liu) [inline examples/ending_locations/subscript.py:20]: now that there's no separating import, ensure there's exactly two blank lines separating the above code with the function definition below

File: python_ta/checkers/forbidden_io_function_checker.py [hunk 2]
```diff
@@ -32,7 +32,7 @@ class IOFunctionChecker(BaseChecker):
                 "default": FORBIDDEN_BUILTIN,
                 "type": "csv",
                 "metavar": "<builtin function names>",
-                "help": "List of built-in function names that should not be used.",
+                "help": "list of built-in function names that should not be used.",
```
Comment (david-yz-liu) [inline python_ta/checkers/forbidden_io_function_checker.py:35]: This shouldn't have been changed (it's not a type annotation)

File: python_ta/checkers/forbidden_python_syntax_checker.py [hunk 3]
```diff
@@ -29,7 +29,7 @@ class ForbiddenPythonSyntaxChecker(BaseChecker):
                 "default": (),
                 "type": "csv",
                 "metavar": "<disallowed-syntax>",
-                "help": "List of Python syntax that are not allowed to be used.",
+                "help": "list of Python syntax that are not allowed to be used.",
```
Comment (david-yz-liu) [inline python_ta/checkers/forbidden_python_syntax_checker.py:32]: See above comment

File: python_ta/checkers/invalid_name_checker.py [hunk 4]
```diff
@@ -14,7 +16,7 @@
 # Bad variable names.
 BAD_NAMES = {"l", "I", "O"}
 
-# Set a limit in name length to keep certain variable names short.
+# set a limit in name length to keep certain variable names short.
```
Comment (david-yz-liu) [inline python_ta/checkers/invalid_name_checker.py:19]: As above comment

File: python_ta/checkers/pycodestyle_checker.py [hunk 5]
```diff
@@ -23,7 +23,7 @@ class PycodestyleChecker(BaseRawFileChecker):
                 "default": (),
                 "type": "csv",
                 "metavar": "<pycodestyle-ignore>",
-                "help": "List of Pycodestyle errors to ignore",
+                "help": "list of Pycodestyle errors to ignore",
```
Comment (david-yz-liu) [inline python_ta/checkers/pycodestyle_checker.py:26]: as above

File: python_ta/contracts/__init__.py [hunk 6]
```diff
@@ -24,17 +26,17 @@
 
 ENABLE_CONTRACT_CHECKING = True
 """
-Set to True to enable contract checking.
+set to True to enable contract checking.
```
Comment (david-yz-liu) [inline python_ta/contracts/__init__.py:29]: as above (and for the other docstrings in this file)

File: python_ta/reporters/templates/jquery-3.7.1.slim.min.js [hunk 7]
```diff
@@ -1,2 +1,2 @@
...
...ave",t||e)}}),ce.each("blur focus focusin focusout resize scroll click dblclick mousedown mouseup mousemove mouseover mouseout mouseenter mouseleave change select submit keydown keypress keyup contextmenu".split(" "),function(e,n){ce.fn[n]=function(e,t){return 0<arguments.length?this.on(n,null,e,t):this.trigger(n)}});var Tt=/^[\s\uFEFF\xA0]+|([^\s\uFEFF\xA0])[\s\uFEFF\xA0]+$/g;ce.proxy=function(e,t){var n,r,i;if("string"==typeof t&&(n=e[t],t=e,e=n),v(e))return r=ae.call(arguments,2),(i=function(){return e.apply(t||this,r.concat(ae.call(arguments)))}).guid=e.guid=e.guid||ce.guid++,i},ce.holdReady=function(e){e?ce.readyWait++:ce.ready(!0)},ce.isArray=Array.isArray,ce.parseJSON=JSON.parse,ce.nodeName=fe,ce.isFunction=v,ce.isWindow=y,ce.camelCase=B,ce.type=x,ce.now=Date.now,ce.isNumeric=function(e){var t=ce.type(e);return("number"===t||"string"===t)&&!isNaN(e-parseFloat(e))},ce.trim=function(e){return null==e?"":(e+"").replace(Tt,"$1")},"function"==typeof define&&define.amd&&define("jquery",[],function(){return ce});var Et=ie.jQuery,kt=ie.$;return ce.noConflict=function(e){return ie.$===ce&&(ie.$=kt),e&&ie.jQuery===ce&&(ie.jQuery=Et),ce},"undefined"==typeof e&&(ie.jQuery=ie.$=ce),ce});
```
Comment (david-yz-liu) [inline python_ta/reporters/templates/jquery-3.7.1.slim.min.js:2]: This file shouldn't have been changed at all

File: python_ta/reporters/templates/script.js [hunk 8]
```diff
@@ -13,7 +13,7 @@ $(".sidebar button").click(function () {
  * https://carlanderson.xyz/how-to-animate-on-height-auto/
  */
 function toggleElement(elem) {
-  const expanded = elem.classList.contains("expanded")
+  const expanded = elem.classlist.contains("expanded")
```
Comment (david-yz-liu) [inline python_ta/reporters/templates/script.js:16]: This file also shouldn't have been changed at all; more generally, only `.py` files (and the documentation files) should be changed. Revert all other changes

File: python_ta/transforms/ExprWrapper.py [hunk 9]
```diff
@@ -1,4 +1,4 @@
-from typing import Dict, List, Optional, Union
+from typing import Optional, Union, dict, list
```
Comment (david-yz-liu) [inline python_ta/transforms/ExprWrapper.py:1]: `dict` and `list` do not need to be imported; but, a `__future__` import does need to be added to the top of this file

File: python_ta/transforms/setendings.py [hunk 10]
```diff
@@ -476,9 +476,9 @@ def end_setter_from_source(source_code, pred, only_consumables=False):
     """
 
     def set_endings_from_source(node):
-        # Tuple nodes have an end_col_offset that includes the end paren,
+        # tuple nodes have an end_col_offset that includes the end paren,
```
Comment (david-yz-liu) [inline python_ta/transforms/setendings.py:479]: Don't change these

File: python_ta/transforms/z3_visitor.py [hunk 11]
```diff
@@ -30,7 +30,7 @@ def set_function_def_z3_constraints(self, node: nodes.FunctionDef):
         for ann, arg in zip(annotations, arguments):
             if ann is None:
                 continue
-            # TODO: what to do about subscripts ex. Set[int], List[Set[int]], ...
+            # TODO: what to do about subscripts ex. Set[int], list[Set[int]], ...
```
Comment (david-yz-liu) [inline python_ta/transforms/z3_visitor.py:33]: It's actually fine to change this, but you should be consistent and also change the `Set`

File: tests/test.pylintrc [hunk 12]
```diff
@@ -5,16 +5,16 @@
 # Default max amount of messages for reporter to display.
 pyta-number-of-messages = 0
 
-# (DEPRECATED: Use output-format option below.) Set the [REPORTS] output-format option instead.
+# (DEPRECATED: Use output-format option below.) set the [REPORTS] output-format option instead.
```
Comment (david-yz-liu) [inline tests/test.pylintrc:8]: No changes in this file

File: tests/test_cfg/test_cfg_generator.py [hunk 13]
```diff
@@ -18,7 +18,7 @@ def reset_dir(request, monkeypatch):
 @pytest.fixture
 def create_cfg():
     """Create the CFG for each test that uses the code in a file fixture."""
-    # Setup: store the paths of the files being used/created
+    # setup: store the paths of the files being used/created
```
Comment (david-yz-liu) [inline tests/test_cfg/test_cfg_generator.py:21]: Don't change these

File: tests/test_config/file_fixtures/test.pylintrc [hunk 14]
```diff
@@ -3,15 +3,15 @@
 # Default max amount of messages for reporter to display.
 pyta-number-of-messages = 10
 
-# Set whether the default error messages by pylint should be overwritten by python TA's custom messages
+# set whether the default error messages by pylint should be overwritten by python TA's custom messages
```
Comment (david-yz-liu) [inline tests/test_config/file_fixtures/test.pylintrc:6]: Don't change this file

File: tests/test_contracts/test_class_forward_reference.py [hunk 15]
```diff
@@ -21,7 +21,7 @@ def __init__(self, item: Any) -> None:
         self.next = None
 
     def bad_method(self) -> None:
-        """Set self.next to an invalid value, violating the type annotation."""
+        """set self.next to an invalid value, violating the type annotation."""
```
Comment (david-yz-liu) [inline tests/test_contracts/test_class_forward_reference.py:24]: Don't change this

File: tests/test_custom_checkers/test_type_annotation_checker.py [hunk 16]
```diff
@@ -12,12 +12,12 @@ def test_type_is_assigned_for_function(self):
         """Ensure that checker catches when type is assigned instead of annotated
         in function parameters."""
         src = """
-        from typing import List
+        from typing import list
```
Comment (david-yz-liu) [inline tests/test_custom_checkers/test_type_annotation_checker.py:15]: Delete this line (and the blank lines below it)

File: tests/test_messages_config/test_no_user_config_no_pyta_overwrite.pylintrc [hunk 17]
```diff
@@ -2,7 +2,7 @@
 
 # Make sure to register custom options tuple first in `python_ta/__init__.py`
 # ===========================================================
-# Set whether the default error messages by pylint should be overwritten by python TA's custom messages
+# set whether the default error messages by pylint should be overwritten by python TA's custom messages
```
Comment (david-yz-liu) [inline tests/test_messages_config/test_no_user_config_no_pyta_overwrite.pylintrc:5]: Don't change this (or the other `.pylintrc` files below)

**Review by david-yz-liu — COMMENTED** [review]
@CulmoneY okay good work, though overall you were a bit too broad with your find-and-replace (see inline comments).

In addition to the comments I left inline, I have two additional comments:

1. You'll need to update your branch since I merged in @Raine-Yang-UofT's PR (see my message on Slack). There will be merge conflicts, so you have a good opportunity to practice resolving those.
2. Update the Changelog. This is an internal change.

**Comment by david-yz-liu** [comment]
@CulmoneY oh sorry I realized you hadn't yet requested a review, so I may have been a bit premature! Well, keep working :)

File: python_ta/contracts/__init__.py [hunk 18]
```diff
@@ -29,12 +31,12 @@
 
 DEBUG_CONTRACTS = False
 """
-Set to True to display debugging messages when checking contracts.
+set to True to display debugging messages when checking contracts.
```
Comment (david-yz-liu) [inline python_ta/contracts/__init__.py:34]: This still needs to be reverted (as below)

File: python_ta/contracts/__init__.py [hunk 19]
```diff
@@ -578,9 +580,9 @@ def _display_annotation(annotation: Any) -> str:
     'int'
     >>> _display_annotation(list[int])
     'list[int]'
-    >>> from typing import List
-    >>> _display_annotation(List[int])
-    'typing.List[int]'
+    >>> from typing import list
```
Comment (david-yz-liu) [inline python_ta/contracts/__init__.py:583]: This entire doctest example needs to be updated (note that you can, and should, run the doctest example manually to ensure it passes)

File: python_ta/transforms/setendings.py [hunk 20]
```diff
@@ -478,7 +478,7 @@ def end_setter_from_source(source_code, pred, only_consumables=False):
     def set_endings_from_source(node):
         # Tuple nodes have an end_col_offset that includes the end paren,
         # but their col_offset does not include the start paren.
-        # To address this, we override the Tuple node's end_col_offset.
+        # To address this, we override the tuple node's end_col_offset.
```
Comment (david-yz-liu) [inline python_ta/transforms/setendings.py:481]: revert this change

File: python_ta/z3/z3_parser.py [hunk 21]
```diff
@@ -178,14 +181,14 @@ def parse_bool_op(self, node: astroid.BoolOp) -> z3.ExprRef:
 
     def parse_container_op(
         self, node: Union[nodes.List, astroid.Set, astroid.Tuple]
-    ) -> List[z3.ExprRef]:
-        """Convert an astroid List, Set, Tuple node to a list of z3 expressions."""
+    ) -> list[z3.ExprRef]:
+        """Convert an astroid list, Set, tuple node to a list of z3 expressions."""
```
Comment (david-yz-liu) [inline python_ta/z3/z3_parser.py:185]: Revert these changes

File: tests/test_cfg/test_cfg_generator.py [hunk 22]
```diff
@@ -43,7 +43,7 @@ def create_cfg_funcs_only():
 
     This fixture specifies that cfgs will only be created for functions.
     """
-    # Setup: store the paths of the files being used/created
+    # setup: store the paths of the files being used/created
```
Comment (david-yz-liu) [inline tests/test_cfg/test_cfg_generator.py:46]: revert this change

File: tests/test_config/file_fixtures/test.pylintrc (hunk 14 above)
Comment (david-yz-liu) [reply tests/test_config/file_fixtures/test.pylintrc:6]: I'm not sure why you marked this comment as resolved; perhaps you meant to make this change, but either forgot or did not commit the change. Either way, please ensure that you double-check your work before resolving my comments, as I don't want to repeat comments.

File: tests/test_contracts/test_class_contracts.py [hunk 23]
```diff
@@ -25,7 +26,7 @@ class Person:
 
     age: int
     name: str
-    fav_foods: List[str]
+    fav_foods: list[str]
```
Comment (david-yz-liu) [inline tests/test_contracts/test_class_contracts.py:29]: Okay so the tests on 3.8 are failing, and not just because of the flaky test we previously discussed. I realized that because 3.8 doesn't support the newer type annotations, we can't change these type annotations *for functions and classes that are used for contract-checking testing. This includes the functions/classes in this file and other files under `tests/test_contracts`. Please revert those changes.

**Comment by CulmoneY** [comment]
I wanted to bring to your attention that the tests in `test_z3_constraints.py` began failing after I merged the latest changes from the main branch. This file was not modified as part of my PR, and I believe the issue might be related to recent updates in the main branch. As such, I am requesting for a PR review. Please let me know if you require any additional steps.

**Review by david-yz-liu — COMMENTED** [review]
@CulmoneY it's great to see that the tests are passing. I left a few more comments, but also it appears you haven't addressed all of the comments I left in my previous review. You should go through each of these carefully; one strategy to keep track of your progress is to use an emoji reaction to indicate when you've fixed (_and committed_) a change for a particular comment.

Don't resolve any comments until you've pushed your changes and verified that the issue is actually fixed (there was at least one instance I saw where you had resolved a comment but hadn't actually made the corresponding change).

File: examples/ending_locations/subscript.py [hunk 24]
```diff
@@ -1,3 +1,4 @@
+from __future__ import annotations
```
[the remark this answers is not in the document]
Comment (CulmoneY) [reply examples/ending_locations/subscript.py:1]: I have made the requested change by adding the blank line as instructed. However, I reviewed the test within test_setendings.py, specifically the section where the test checks the line numbers of subscripts through AST traversal. As a result, I modified the relevant lines in the test to reflect the correct line numbers of the subscript nodes.

File: tests/test_custom_checkers/test_type_annotation_checker.py (hunk 16 above)
Comment (CulmoneY) [reply tests/test_custom_checkers/test_type_annotation_checker.py]: This change leads to a failing test in Python 3.8. As such, it has been reverted.

File: python_ta/contracts/__init__.py (hunk 19 above)
Comment (CulmoneY) [reply python_ta/contracts/__init__.py:583]: from my understanding, the doctest is designed to demonstrate three different types of annotations, including one that uses from typing import List. Since the class already fails in Python 3.8 due to the list[int] syntax producing a type error, I have reverted my previous changes to avoid further issues.

**Comment by CulmoneY** [comment]
Hi David,

All the requested changes have been implemented. I double-checked everything to ensure that there are no unnecessary or incomplete modifications. However, I would like to apologize for any inconvenience caused by premature reviews. I converted my PR from draft status to run the tests, which is why some of your requested corrections weren't implemented at the time.

Moving forward, I’ll be more mindful to avoid this and ensure that my PR is fully ready before requesting a review.

Thank you for your understanding, and I appreciate your time and feedback.

**Review by david-yz-liu — COMMENTED** [review]
@CulmoneY looks great! For future reference, you can unmark a PR as draft in order to trigger the CI tests, and that's a separate actions from "requesting a review" from me (the latter is a separate action on GitHub). This lets you run the CI tests without yet asking for me to review it.
